File: src/pdfjet/textline.go
package pdfjet

/**
 * textline.go
 *
Copyright 2020 Innovatics Inc.

Redistribution and use in source and binary forms, with or without modification,
are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice,
  this list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and / or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

import (
	"color"
	"effect"
	"math"
)

// TextLine is used to create text line objects.
type TextLine struct {
	text               string
	x, y, xBox, yBox   float32
	font, fallbackFont *Font
	trailingSpace      bool
	uri, key           *string
	underline          bool
	strikeout          bool
	underlineTTS       string
	strikeoutTTS       string
	degrees            int
	color              uint32
	textEffect         int
	verticalOffset     float32
	language           string
	altDescription     string
	actualText         string
	uriLanguage        string
	uriAltDescription  string
	uriActualText      string
}

// NewTextLine is constructor for creating text line objects.
// @param font the font to use.
// @param text the text.
func NewTextLine(font *Font, text string) *TextLine {
	textLine := new(TextLine)
	textLine.font = font
	textLine.text = text
	textLine.trailingSpace = true
	textLine.underlineTTS = "underline"
	textLine.strikeoutTTS = "strikeout"
	textLine.color = color.Black
	textLine.textEffect = effect.Normal
	textLine.verticalOffset = 0.0
	textLine.altDescription = text
	textLine.actualText = text
	return textLine
}

// SetText sets the text.
// @param text the text.
// @return this TextLine.
func (textLine *TextLine) SetText(text string) *TextLine {
	textLine.text = text
	if textLine.altDescription == "" {
		textLine.altDescription = text
	}
	if textLine.actualText == "" {
		textLine.actualText = text
	}
	return textLine
}

// GetText returns the text.
func (textLine *TextLine) GetText() string {
	return textLine.text
}

// SetLocation sets the location where this text line will be drawn on the page.
// @param x the x coordinate of the text line.
// @param y the y coordinate of the text line.
// @return this TextLine.
func (textLine *TextLine) SetLocation(x, y float32) *TextLine {
	textLine.x = x
	textLine.y = y
	return textLine
}

// SetFont sets the font to use for this text line.
// @param font the font to use.
// @return this TextLine.
func (textLine *TextLine) SetFont(font *Font) *TextLine {
	textLine.font = font
	return textLine
}

// GetFont gets the font to use for this text line.
// @return font the font to use.
func (textLine *TextLine) GetFont() *Font {
	return textLine.font
}

// SetFontSize sets the font size to use for this text line.
// @param fontSize the fontSize to use.
// @return this TextLine.
func (textLine *TextLine) SetFontSize(fontSize float32) *TextLine {
	textLine.font.SetSize(fontSize)
	return textLine
}

// SetFallbackFont sets the fallback font.
// @param fallbackFont the fallback font.
// @return this TextLine.
func (textLine *TextLine) SetFallbackFont(fallbackFont *Font) *TextLine {
	textLine.fallbackFont = fallbackFont
	return textLine
}

// SetFallbackFontSize sets the fallback font size to use for this text line.
// @param fallbackFontSize the fallback font size.
// @return this TextLine.
func (textLine *TextLine) SetFallbackFontSize(fallbackFontSize float32) *TextLine {
	textLine.fallbackFont.SetSize(fallbackFontSize)
	return textLine
}

// GetFallbackFont returns the fallback font.
// @return the fallback font.
func (textLine *TextLine) GetFallbackFont() *Font {
	return textLine.fallbackFont
}

// SetColor sets the color for this text line.
// @param color the color is specified as an integer.
// @return this TextLine.
func (textLine *TextLine) SetColor(color uint32) *TextLine {
	textLine.color = color
	return textLine
}

// SetColorRGB sets the pen color.
// @param color the color. See the Color class for predefined values or define your own using 0x00RRGGBB packed integers.
// @return this TextLine.
func (textLine *TextLine) SetColorRGB(color []uint32) *TextLine {
	textLine.color = color[0]<<16 | color[1]<<8 | color[2]
	return textLine
}

// GetColor returns the text line color.
// @return the text line color.
func (textLine *TextLine) GetColor() uint32 {
	return textLine.color
}

// GetDestinationY returns the y coordinate of the destination.
// @return the y coordinate of the destination.
func (textLine *TextLine) GetDestinationY() float32 {
	return textLine.y - textLine.font.GetSize()
}

// GetWidth returns the width of this TextLine.
// @return the width.
func (textLine *TextLine) GetWidth() float32 {
	return textLine.font.StringWidth(textLine.fallbackFont, textLine.text)
}

// GetHeight returns the height of this TextLine.
// @return the height.
func (textLine *TextLine) GetHeight() float32 {
	return textLine.font.GetHeight()
}

// SetURIAction sets the URI for the "click text line" action.
// @param uri the URI
// @return this TextLine.
func (textLine *TextLine) SetURIAction(uri *string) *TextLine {
	textLine.uri = uri
	return textLine
}

// GetURIAction returns the action URI.
// @return the action URI.
func (textLine *TextLine) GetURIAction() *string {
	return textLine.uri
}

// SetGoToAction sets the destination key for the action.
// @param key the destination name.
// @return this TextLine.
func (textLine *TextLine) SetGoToAction(key *string) *TextLine {
	textLine.key = key
	return textLine
}

// GetGoToAction returns the GoTo action string.
// @return the GoTo action string.
func (textLine *TextLine) GetGoToAction() *string {
	return textLine.key
}

// SetUnderline sets the underline variable.
// If the value of the underline variable is 'true' - the text is underlined.
// @param underline the underline flag.
// @return this TextLine.
func (textLine *TextLine) SetUnderline(underline bool) *TextLine {
	textLine.underline = underline
	return textLine
}

// GetUnderline returns the underline flag.
// @return the underline flag.
func (textLine *TextLine) GetUnderline() bool {
	return textLine.underline
}

// SetStrikeout sets the strike variable.
// If the value of the strike variable is 'true' - a strike line is drawn through the text.
// @param strikeout the strikeout flag.
// @return this TextLine.
func (textLine *TextLine) SetStrikeout(strikeout bool) *TextLine {
	textLine.strikeout = strikeout
	return textLine
}

// GetStrikeout returns the strikeout flag.
// @return the strikeout flag.
func (textLine *TextLine) GetStrikeout() bool {
	return textLine.strikeout
}

// SetTextDirection sets the direction in which to draw the text.
// @param degrees the number of degrees.
// @return this TextLine.
func (textLine *TextLine) SetTextDirection(degrees int) *TextLine {
	textLine.degrees = degrees
	return textLine
}

// GetTextDirection returns the text direction.
// @return the text direction.
func (textLine *TextLine) GetTextDirection() int {
	return textLine.degrees
}

// SetTextEffect sets the text effect.
// @param textEffect Effect.NORMAL, Effect.SUBSCRIPT or Effect.SUPERSCRIPT.
// @return this TextLine.
func (textLine *TextLine) SetTextEffect(textEffect int) *TextLine {
	textLine.textEffect = textEffect
	return textLine
}

// GetTextEffect returns the text effect.
// @return the text effect.
func (textLine *TextLine) GetTextEffect() int {
	return textLine.textEffect
}

// SetVerticalOffset sets the vertical offset of the text.
// @param verticalOffset the vertical offset.
// @return this TextLine.
func (textLine *TextLine) SetVerticalOffset(verticalOffset float32) *TextLine {
	textLine.verticalOffset = verticalOffset
	return textLine
}

// GetVerticalOffset returns the vertical text offset.
// @return the vertical text offset.
func (textLine *TextLine) GetVerticalOffset() float32 {
	return textLine.verticalOffset
}

// SetTrailingSpace sets the trailing space after this text line when used in paragraph.
// @param trailingSpace the trailing space.
// @return this TextLine.
func (textLine *TextLine) SetTrailingSpace(trailingSpace bool) *TextLine {
	textLine.trailingSpace = trailingSpace
	return textLine
}

// GetTrailingSpace returns the trailing space.
// @return the trailing space.
func (textLine *TextLine) GetTrailingSpace() bool {
	return textLine.trailingSpace
}

// SetLanguage sets the language.
func (textLine *TextLine) SetLanguage(language string) *TextLine {
	textLine.language = language
	return textLine
}

// GetLanguage gets the language.
func (textLine *TextLine) GetLanguage() string {
	return textLine.language
}

// SetAltDescription sets the alternate description of this text line.
// @param altDescription the alternate description of the text line.
// @return this TextLine.
func (textLine *TextLine) SetAltDescription(altDescription string) *TextLine {
	textLine.altDescription = altDescription
	return textLine
}

// GetAltDescription gets the alternate description of this text line.
func (textLine *TextLine) GetAltDescription() string {
	return textLine.altDescription
}

// SetActualText sets the actual text for this text line.
// @param actualText the actual text for the text line.
// @return this TextLine.
func (textLine *TextLine) SetActualText(actualText string) *TextLine {
	textLine.actualText = actualText
	return textLine
}

// GetActualText gets the actual text.
func (textLine *TextLine) GetActualText() string {
	return textLine.actualText
}

// SetURILanguage sets the URI language.
func (textLine *TextLine) SetURILanguage(uriLanguage string) *TextLine {
	textLine.uriLanguage = uriLanguage
	return textLine
}

// SetURIAltDescription sets the URI alternative description.
func (textLine *TextLine) SetURIAltDescription(uriAltDescription string) *TextLine {
	textLine.uriAltDescription = uriAltDescription
	return textLine
}

// SetURIActualText sets the URI actual text.
func (textLine *TextLine) SetURIActualText(uriActualText string) *TextLine {
	textLine.uriActualText = uriActualText
	return textLine
}

// PlaceInAtZeroZero places this text line in the specified box at location (0.0, 0.0)
func (textLine *TextLine) PlaceInAtZeroZero(box *Box) *TextLine {
	textLine.PlaceIn(box, 0.0, 0.0)
	return textLine
}

// PlaceIn places this text line in the box at the specified offset.
// @param box the specified box.
// @param xOffset the x offset from the top left corner of the box.
// @param yOffset the y offset from the top left corner of the box.
// @return this TextLine.
func (textLine *TextLine) PlaceIn(box *Box, xOffset, yOffset float32) *TextLine {
	textLine.xBox = box.x + xOffset
	textLine.yBox = box.y + yOffset
	return textLine
}

// DrawOn draws this text line on the specified page if the draw parameter is true.
// @param page the page to draw this text line on.
// @param draw if draw is false - no action is performed.
func (textLine *TextLine) DrawOn(page *Page) []float32 {
	if page == nil || textLine.text == "" {
		return []float32{textLine.x, textLine.y}
	}

	page.SetTextDirection(textLine.degrees)

	textLine.x += textLine.xBox
	textLine.y += textLine.yBox

	page.SetBrushColor(textLine.color)
	page.AddBMC("Span", textLine.language, textLine.altDescription, textLine.actualText)
	page.DrawString(textLine.font, textLine.fallbackFont, textLine.text, textLine.x, textLine.y)
	page.AddEMC()

	radians := float64(math.Pi) * float64(textLine.degrees) / float64(180.0)
	if textLine.underline {
		page.SetPenWidth(textLine.font.underlineThickness)
		page.SetPenColor(textLine.color)
		lineLength := textLine.font.StringWidth(textLine.fallbackFont, textLine.text)
		xAdjust := textLine.font.underlinePosition*float32(math.Sin(radians)) + textLine.verticalOffset
		yAdjust := textLine.font.underlinePosition*float32(math.Cos(radians)) + textLine.verticalOffset
		x2 := textLine.x + lineLength*float32(math.Cos(radians))
		y2 := textLine.y - lineLength*float32(math.Sin(radians))
		page.AddBMC("Span", textLine.language, textLine.underlineTTS, textLine.underlineTTS)
		page.MoveTo(textLine.x+xAdjust, textLine.y+yAdjust)
		page.LineTo(x2+xAdjust, y2+yAdjust)
		page.StrokePath()
		page.AddEMC()
	}

	if textLine.strikeout {
		page.SetPenWidth(textLine.font.underlineThickness)
		page.SetPenColor(textLine.color)
		lineLength := textLine.font.StringWidth(textLine.fallbackFont, textLine.text)
		xAdjust := (textLine.font.bodyHeight / 4.0) * float32(math.Sin(radians))
		yAdjust := (textLine.font.bodyHeight / 4.0) * float32(math.Cos(radians))
		x2 := textLine.x + lineLength*float32(math.Cos(radians))
		y2 := textLine.y - lineLength*float32(math.Sin(radians))
		page.AddBMC("Span", textLine.language, textLine.strikeoutTTS, textLine.strikeoutTTS)
		page.MoveTo(textLine.x-xAdjust, textLine.y-yAdjust)
		page.LineTo(x2-xAdjust, y2-yAdjust)
		page.StrokePath()
		page.AddEMC()
	}

	if textLine.uri != nil || textLine.key != nil {
		page.AddAnnotation(NewAnnotation(
			textLine.uri,
			textLine.key, // The destination name
			textLine.x,
			textLine.y-textLine.font.ascent,
			textLine.x+textLine.font.StringWidth(textLine.fallbackFont, textLine.text),
			textLine.y+textLine.font.descent,
			textLine.uriLanguage,
			textLine.uriAltDescription,
			textLine.uriActualText))
	}

	page.SetTextDirection(0)

	length := textLine.font.StringWidth(textLine.fallbackFont, textLine.text)
	xMax := math.Max(float64(textLine.x), float64(textLine.x)+float64(length)*math.Cos(radians))
	yMax := math.Max(float64(textLine.y), float64(textLine.y)-float64(length)*math.Sin(radians))

	return []float32{float32(xMax), float32(yMax)}
}

func (textLine *TextLine) advance(leading float32) float32 {
	textLine.y += leading
	return textLine.y
}

File: Sources/Sources/PDFjet/TextLine.swift
/**
 *  TextLine.swift
 *
Copyright 2020 Innovatics Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
import Foundation


///
/// Used to create text line objects.
///
public class TextLine : Drawable {

    var x: Float = 0.0
    var y: Float = 0.0

    var font: Font?
    var fallbackFont: Font?
    var text: String?
    var trailingSpace: Bool = true

    private var uri: String?
    private var key: String?

    private var underline = false
    private var strikeout = false

    private var degrees = 0
    private var color = Color.black

    private var xBox: Float = 0.0
    private var yBox: Float = 0.0

    private var textEffect = Effect.NORMAL
    private var verticalOffset: Float = 0.0

    private var language: String?
    private var altDescription: String?

    private var uriLanguage: String?
    private var uriActualText: String?
    private var uriAltDescription: String?

    private var structureType = StructElem.P


    ///
    /// Constructor for creating text line objects.
    ///
    /// - Parameter font the font to use.
    ///
    public init(_ font: Font) {
        self.font = font
    }


    ///
    /// Constructor for creating text line objects.
    ///
    /// - Parameter font the font to use.
    /// - Parameter text the text.
    ///
    public init(_ font: Font, _ text: String) {
        self.font = font
        self.text = text
        self.altDescription = text
    }


    ///
    /// Sets the text.
    ///
    /// - Parameter text the text.
    /// - Returns: selt.TextLine.
    ///
    @discardableResult
    public func setText(_ text: String) -> TextLine {
        self.text = text
        self.altDescription = text
        return self
    }


    ///
    /// Returns the text.
    ///
    /// - Returns: the text.
    ///
    public func getText() -> String? {
        return self.text
    }


    public func setPosition(_ x: Float, _ y: Float) {
        setLocation(x, y)
    }


    ///
    /// Sets the location where selt.text line will be drawn on the page.
    ///
    /// - Parameter x the x coordinate of the text line.
    /// - Parameter y the y coordinate of the text line.
    /// - Returns: selt.TextLine.
    ///
    @discardableResult
    public func setLocation(_ x: Float, _ y: Float) -> TextLine {
        self.x = x;
        self.y = y;
        return self;
    }


    ///
    /// Sets the font to use for selt.text line.
    ///
    /// - Parameter font the font to use.
    /// - Returns: selt.TextLine.
    ///
    @discardableResult
    public func setFont(_ font: Font) -> TextLine {
        self.font = font
        return self
    }


    ///
    /// Gets the font to use for selt.text line.
    ///
    /// - Returns: font the font to use.
    ///
    public func getFont() -> Font {
        return self.font!
    }


    ///
    /// Sets the font size to use for selt.text line.
    ///
    /// - Parameter fontSize the fontSize to use.
    /// - Returns: selt.TextLine.
    ///
    @discardableResult
    public func setFontSize(_ fontSize: Float) -> TextLine {
        self.font!.setSize(fontSize)
        return self
    }


    ///
    /// Sets the fallback font.
    ///
    /// - Parameter fallbackFont the fallback font.
    /// - Returns: selt.TextLine.
    ///
    @discardableResult
    public func setFallbackFont(_ fallbackFont: Font?) -> TextLine {
        self.fallbackFont = fallbackFont
        return self
    }


    ///
    /// Sets the fallback font size to use for selt.text line.
    ///
    /// - Parameter fallbackFontSize the fallback font size.
    /// - Returns: selt.TextLine.
    ///
    @discardableResult
    public func setFallbackFontSize(_ fallbackFontSize: Float) -> TextLine {
        self.fallbackFont!.setSize(fallbackFontSize)
        return self
    }


    ///
    /// Returns the fallback font.
    ///
    /// - Returns: the fallback font.
    ///
    public func getFallbackFont() -> Font? {
        return self.fallbackFont
    }


    ///
    /// Sets the color for selt.text line.
    ///
    /// - Parameter color the color is specified as an integer.
    /// - Returns: selt.TextLine.
    ///
    @discardableResult
    public func setColor(_ color: UInt32) -> TextLine {
        self.color = color
        return self
    }


    ///
    /// Sets the pen color.
    ///
    /// - Parameter color the color.
    ///   See the Color class for predefined values or define your own using 0x00RRGGBB packed integers.
    /// - Returns: selt.TextLine.
    ///
    @discardableResult
    public func setColor(_ color: [UInt32]) -> TextLine {
        self.color = color[0] << 16 | color[1] << 8 | color[2]
        return self
    }


    ///
    /// Returns the text line color.
    ///
    /// - Returns: the text line color.
    ///
    public func getColor() -> UInt32 {
        return self.color
    }


    ///
    /// Returns the y coordinate of the destination.
    ///
    /// - Returns: the y coordinate of the destination.
    ///
    public func getDestinationY() -> Float {
        return y - font!.getSize()
    }


    ///
    /// Returns the width of selt.TextLine.
    ///
    /// - Returns: the width.
    ///
    public func getWidth() -> Float {
        return font!.stringWidth(fallbackFont, text!)
    }


    public func getStringWidth(_ text: String) -> Float {
        return font!.stringWidth(fallbackFont, text)
    }


    ///
    /// Returns the height of selt.TextLine.
    ///
    /// - Returns: the height.
    ///
    public func getHeight() -> Float {
        return font!.getHeight()
    }


    ///
    /// Sets the URI for the "click text line" action.
    ///
    /// - Parameter uri the URI
    /// - Returns: selt.TextLine.
    ///
    @discardableResult
    public func setURIAction(_ uri: String?) -> TextLine {
        self.uri = uri
        return self
    }


    ///
    /// Returns the action URI.
    ///
    /// - Returns: the action URI.
    ///
    public func getURIAction() -> String? {
        return self.uri
    }


    ///
    /// Sets the destination key for the action.
    ///
    /// - Parameter key the destination name.
    /// - Returns: selt.TextLine.
    ///
    @discardableResult
    public func setGoToAction(_ key: String?) -> TextLine {
        self.key = key
        return self
    }


    ///
    /// Returns the GoTo action string.
    ///
    /// - Returns: the GoTo action string.
    ///
    public func getGoToAction() -> String? {
        return self.key
    }


    ///
    /// Sets the underline variable.
    /// If the value of the underline variable is 'true' - the text is underlined.
    ///
    /// - Parameter underline the underline flag.
    /// - Returns: selt.TextLine.
    ///
    @discardableResult
    public func setUnderline(_ underline: Bool) -> TextLine {
        self.underline = underline
        return self
    }


    ///
    /// Returns the underline flag.
    ///
    /// - Returns: the underline flag.
    ///
    public func getUnderline() -> Bool {
        return self.underline
    }


    ///
    /// Sets the strike variable.
    /// If the value of the strike variable is 'true' - a strike line is drawn through the text.
    ///
    /// - Parameter strikeout the strikeout flag.
    /// - Returns: selt.TextLine.
    ///
    @discardableResult
    public func setStrikeout(_ strikeout: Bool) -> TextLine {
        self.strikeout = strikeout
        return self
    }


    ///
    /// Returns the strikeout flag.
    ///
    /// - Returns: the strikeout flag.
    ///
    public func getStrikeout() -> Bool {
        return self.strikeout
    }


    ///
    /// Sets the direction in which to draw the text.
    ///
    /// - Parameter degrees the number of degrees.
    /// - Returns: selt.TextLine.
    ///
    @discardableResult
    public func setTextDirection(_ degrees: Int) -> TextLine {
        self.degrees = degrees
        return self
    }


    ///
    /// Returns the text direction.
    ///
    /// - Returns: the text direction.
    ///
    public func getTextDirection() -> Int {
        return degrees
    }


    ///
    /// Sets the text effect.
    ///
    /// - Parameter textEffect Effect.NORMAL, Effect.SUBSCRIPT or Effect.SUPERSCRIPT.
    /// - Returns: selt.TextLine.
    ///
    @discardableResult
    public func setTextEffect(_ textEffect: Int) -> TextLine {
        self.textEffect = textEffect
        return self
    }


    ///
    /// Returns the text effect.
    ///
    /// - Returns: the text effect.
    ///
    public func getTextEffect() -> Int {
        return self.textEffect
    }


    ///
    /// Sets the vertical offset of the text.
    ///
    /// - Parameter verticalOffset the vertical offset.
    /// - Returns: selt.TextLine.
    ///
    @discardableResult
    public func setVerticalOffset(_ verticalOffset: Float) -> TextLine {
        self.verticalOffset = verticalOffset
        return self
    }


    ///
    /// Returns the vertical text offset.
    ///
    /// - Returns: the vertical text offset.
    ///
    public func getVerticalOffset() -> Float {
        return self.verticalOffset
    }


    ///
    /// Sets the trailing space after selt.text line when used in paragraph.
    ///
    /// - Parameter trailingSpace the trailing space.
    /// - Returns: selt.TextLine.
    ///
    @discardableResult
    public func setTrailingSpace(_ trailingSpace: Bool) -> TextLine {
        self.trailingSpace = trailingSpace
        return self
    }


    ///
    /// Returns the trailing space.
    ///
    /// - Returns: the trailing space.
    ///
    public func getTrailingSpace() -> Bool {
        return self.trailingSpace
    }


    @discardableResult
    public func setLanguage(_ language: String?) -> TextLine {
        self.language = language
        return self
    }


    public func getLanguage() -> String? {
        return self.language
    }


    ///
    /// Sets the alternate description of selt.text line.
    ///
    /// - Parameter altDescription the alternate description of the text line.
    /// - Returns: selt.TextLine.
    ///
    @discardableResult
    public func setAltDescription(_ altDescription: String?) -> TextLine {
        self.altDescription = altDescription
        return self
    }


    public func getAltDescription() -> String? {
        return self.altDescription
    }


    @discardableResult
    public func setURILanguage(_ uriLanguage: String?) -> TextLine {
        self.uriLanguage = uriLanguage
        return self
    }


    @discardableResult
    public func setURIAltDescription(_ uriAltDescription: String?) -> TextLine {
        self.uriAltDescription = uriAltDescription
        return self
    }


    @discardableResult
    public func setURIActualText(_ uriActualText: String?) -> TextLine {
        self.uriActualText = uriActualText
        return self
    }


    @discardableResult
    public func setStructureType(_ structureType: String) -> TextLine {
        self.structureType = structureType
        return self
    }


    ///
    /// Places selt.text line in the specified box.
    ///
    /// - Parameter box the specified box.
    /// - Returns: selt.TextLine.
    ///
    @discardableResult
    public func placeIn(_ box: Box) -> TextLine {
        placeIn(box, 0.0, 0.0)
        return self
    }


    ///
    /// Places selt.text line in the box at the specified offset.
    ///
    /// - Parameter box the specified box.
    /// - Parameter xOffset the x offset from the top left corner of the box.
    /// - Parameter yOffset the y offset from the top left corner of the box.
    /// - Returns: selt.TextLine.
    ///
    @discardableResult
    public func placeIn(
            _ box: Box,
            _ xOffset: Float,
            _ yOffset: Float) -> TextLine {
        self.xBox = box.x + xOffset
        self.yBox = box.y + yOffset
        return self
    }


    public func advance(_ leading: Float) -> Float {
        self.y += leading
        return self.y
    }


    public func getTextY() -> Float {
        return self.y
    }


    ///
    /// Draws selt.text line on the specified page if the draw parameter is true.
    ///
    /// - Parameter page the page to draw selt.text line on.
    /// - Parameter draw if draw is false - no action is performed.
    ///
    @discardableResult
    public func drawOn(_ page: Page?) -> [Float] {
        if text == nil || text == "" {
            return [x, y]
        }

        page!.setTextDirection(degrees)

        self.x += xBox
        self.y += yBox

        page!.setBrushColor(color)
        page!.addBMC(structureType, language, text!, altDescription!)
        page!.drawString(font!, fallbackFont, text, self.x, self.y)
        page!.addEMC()

        let radians = Float.pi * Float(degrees) / 180.0
        if underline {
            page!.setPenWidth(font!.underlineThickness)
            page!.setPenColor(color)
            let lineLength = font!.stringWidth(fallbackFont, text!)
            let xAdjust = font!.underlinePosition * Float(sin(radians)) + verticalOffset
            let yAdjust = font!.underlinePosition * Float(cos(radians)) + verticalOffset
            let x2 = x + lineLength * Float(cos(radians))
            let y2 = y - lineLength * Float(sin(radians))
            page!.addBMC(structureType, language, text!, "Underlined text: " + text!)
            page!.moveTo(x + xAdjust, y + yAdjust)
            page!.lineTo(x2 + xAdjust, y2 + yAdjust)
            page!.strokePath()
            page!.addEMC()
        }

        if strikeout {
            page!.setPenWidth(font!.underlineThickness)
            page!.setPenColor(color)
            let lineLength = font!.stringWidth(fallbackFont, text!)
            let xAdjust = (font!.bodyHeight / 4.0) * Float(sin(radians))
            let yAdjust = (font!.bodyHeight / 4.0) * Float(cos(radians))
            let x2 = x + lineLength * Float(cos(radians))
            let y2 = y - lineLength * Float(sin(radians))
            page!.addBMC(structureType, language, text!, "Strikethrough text: " + text!)
            page!.moveTo(x - xAdjust, y - yAdjust)
            page!.lineTo(x2 - xAdjust, y2 - yAdjust)
            page!.strokePath()
            page!.addEMC()
        }

        if uri != nil || key != nil {
            page!.addAnnotation(Annotation(
                    uri,
                    key,    // The destination name
                    self.x,
                    self.y - font!.ascent,
                    self.x + font!.stringWidth(fallbackFont, text!),
                    self.y + font!.descent,
                    uriLanguage,
                    uriActualText,
                    uriAltDescription))
        }
        page!.setTextDirection(0)

        let len = font!.stringWidth(fallbackFont, text!)
        let xMax = max(x, x + len*Float(cos(radians)))
        let yMax = max(y, y - len*Float(sin(radians)))

        return [xMax, yMax]
    }

}   // End of TextLine.swift
